jobserver/pipeline_config.py

```diff
@@ -1,28 +1,24 @@
-import yaml
 from pygments import highlight
 from pygments.formatters import HtmlFormatter
 from pygments.lexers import YamlLexer
 
 from .github import get_branch, get_file
 
 
-def get_actions(project, status_lut):
-    """Get actions from project.yaml for this Workspace"""
-    # ensure there's always a run_all action
-    if "run_all" not in project["actions"]:
-        project["actions"]["run_all"] = {"needs": list(project["actions"].keys())}
-
-    for action, children in project["actions"].items():
-        needs = children.get("needs", []) if children else []
-
-        # handle needs being defined but empty
-        if needs is None:
-            needs = []
+def get_actions(config, status_lut):
+    """Get actions from a pipeline config for this Workspace"""
+    for action, children in config.actions.items():
+        needs = sorted(children.needs)
 
         # get latest status for this action from the lookup table
         status = status_lut.get(action, "-")
 
-        yield {"name": action, "needs": sorted(needs), "status": status}
+        yield {"name": action, "needs": needs, "status": status}
+
+    # ensure there's always a run_all action
+    if "run_all" not in config.actions:
+        all_actions = list(config.actions.keys())
+        yield {"name": "run_all", "needs": all_actions, "status": "-"}
```

Comment:
I'm sure this has been discussed elsewhere, but why must there always be a `run_all`?

Reply:
This is a funny quirk of OpenSAFELY history!

Originally we asked researchers to manually add a `run_all` action to their `project.yaml`s, there was nothing special about this, they simply had to list all other actions in the `needs` section and it ran them all.  Beautifully hacky and it worked quite well!

Of course this fell apart with the larger `project.yaml`s when there were so many dependencies that human error started to catch people out and it was a real chore to create that action.  Around that same time job-server was being set up and we added the create job page (`JobRequestCreate`) which renders the actions in a `project.yaml`.  This function drives the form construction and validation which deals with "what actions can a user run" as defined in the `project.yaml`, and we added a "fake" `run_all` action for everyone to avoid the manual hack.

job-runner now treats `run_all` as a special case since it might not exist in the `project.yaml` either.

Maybe one day we'll derive this from the create job form page with UI controls and ditch the magic string, but backwards compatibility of studies means we're tied to at least support it in job-runner forever more.